### PR TITLE
[doc][tiered storage] read data from filesystem

### PR DESCRIPTION
**Pull request by liangyepianzhou**
Fix https://github.com/apache/pulsar/issues/14526
### Motivation & Modification
The current Filesystem Offload documentation has no documentation on how to directly read the data after offload, so consider adding documentation for this section

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API: (yes / no)
  - The schema: (yes / no / don't know)
  - The default values of configurations: (yes / no)
  - The wire protocol: (yes / no)
  - The rest endpoints: (yes / no)
  - The admin cli options: (yes / no)
  - Anything that affects deployment: (yes / no / don't know)

### Documentation

Check the box below or label this PR directly.

Need to update docs? 

- [ ] `doc-required` 
(Your PR needs to update docs and you will update later)
  
- [ ] `no-need-doc` 
(Please explain why)
  
- [x] `doc` 
(Your PR contains doc changes)

- [ ] `doc-added`
(Docs have been already added)

### CODE REVIEW

**Comment by Anonymitaet** [comment]
@momo-jun could you please help review this PR? Thanks

**Comment by liangyepianzhou** [comment]
@zymap Could you please give a review for this doc.

File: site2/docs/tiered-storage-filesystem.md [hunk 1]
```diff
@@ -520,4 +520,40 @@ Execute the following commands in the repository where you download Pulsar tarba
...
+
+## Read data from fileSystem
+
+This section provides detailed instructions about how to read data out as Ledger Entry in the file system.
+
+* The data was offloaded as `MapFile` to the following path:
+  ```properties
+    path = storageBasePath + "/" + managedLedgerName + "/" + ledgerId + "-" + uuid.toString();
+  ```
+    1. storageBasePath is the value of `hadoop.tmp.dir`, configured in `broker.conf` and `filesystem_offload_core_site.xml`
```
Comment (congbobo184) [inline site2/docs/tiered-storage-filesystem.md:533]:
```suggestion
    1. storageBasePath is the value of `hadoop.tmp.dir`, configured in `broker.conf` or `filesystem_offload_core_site.xml`
```

File: site2/docs/tiered-storage-filesystem.md [hunk 2]
```diff
@@ -520,4 +520,40 @@ Execute the following commands in the repository where you download Pulsar tarba
...
+    2. managedLedgerName is the name of the persistentTopic manager Ledger
+  ```shell
+     managedLedgerName of persistent://public/default/topics-name is  public/default/persistent/topics-name.
```
Comment (congbobo184) [inline site2/docs/tiered-storage-filesystem.md:536]:
```suggestion
     managedLedgerName of persistent://public/default/topics-name is public/default/persistent/topics-name.
```

File: site2/docs/tiered-storage-filesystem.md [hunk 3]
```diff
@@ -520,4 +520,40 @@ Execute the following commands in the repository where you download Pulsar tarba
...
+  ```
+  Considering the iteration of versions,  you can use the following method to get the managedLedgerName:
```
Comment (congbobo184) [inline site2/docs/tiered-storage-filesystem.md:538]:
```suggestion
  Can use the following method to get the managedLedgerName:
```

File: site2/docs/tiered-storage-filesystem.md [hunk 4]
```diff
@@ -520,4 +520,40 @@ Execute the following commands in the repository where you download Pulsar tarba
...
+     LongWritable key = new LongWritable();
+     BytesWritable value = new BytesWritable();
+     key.set(nextExpectedId - 1);
+     reader.seek(key);
+     reader.next(key, value);
+     int length = value.getLength();
+     long entryId = key.get();
+     ByteBuf buf = PooledByteBufAllocator.DEFAULT.buffer(length, length);
+     buf.writeBytes(value.copyBytes());
+     LedgerEntryImpl ledgerEntry = LedgerEntryImpl.create(ledgerId, entryId, length, buf);
```
Comment (zymap) [inline site2/docs/tiered-storage-filesystem.md:558]: Just want to know what users want to do when they read the entry? The entry is a serialized message, does they will deserialize the entry to the message by themself?

Comment (liangyepianzhou) [reply site2/docs/tiered-storage-filesystem.md:558]: Great advice. I have added the deserialization steps, please review again when you have time.

File: site2/docs/tiered-storage-filesystem.md [hunk 5]
```diff
@@ -520,4 +520,66 @@ Execute the following commands in the repository where you download Pulsar tarba
 
     And the **Capacity Used** is changed from 4 KB to 116.46 KB.
 
-    ![](assets/FileSystem-8.png)
\ No newline at end of file
+    ![](assets/FileSystem-8.png)
+
+## Read data from fileSystem
```
Comment (codelipenghui) [inline site2/docs/tiered-storage-filesystem.md:525]:
```suggestion
## Read offloaded data from fileSystem
```

**Review by zymap — APPROVED** [review]
LGTM

**Review by codelipenghui — APPROVED** [review]

**Comment by codelipenghui** [comment]
@Anonymitaet Please help review this PR.

File: site2/docs/tiered-storage-filesystem.md [hunk 6]
```diff
@@ -520,4 +520,66 @@ Execute the following commands in the repository where you download Pulsar tarba
...
 
-    ![](assets/FileSystem-8.png)
\ No newline at end of file
+    ![](assets/FileSystem-8.png)
+
+## Read offloaded data from fileSystem
+
+This section provides detailed instructions about how to read data out as Ledger Entry in the file system.
+
+* The data was offloaded as `MapFile` to the following path:
```
Comment (momo-jun) [inline site2/docs/tiered-storage-filesystem.md:529]:
```suggestion
The data is offloaded to the following new path as `MapFile`:
```

File: site2/docs/tiered-storage-filesystem.md [hunk 7]
```diff
@@ -520,4 +520,66 @@ Execute the following commands in the repository where you download Pulsar tarba
...
+    2. managedLedgerName is the name of the persistentTopic manager Ledger
+  ```shell
+     managedLedgerName of persistent://public/default/topics-name is public/default/persistent/topics-name.
+  ```
+  Can use the following method to get the managedLedgerName:
+  ```shell
+     String managedLedgerName = TopicName.get("persistent://public/default/topics-name").getPersistenceNamingEncoding(); 
+  ```
+
+* Create a reader to read `MapFile` according to the above path and the `configuration` of the file system
```
Comment (momo-jun) [inline site2/docs/tiered-storage-filesystem.md:543]:
```suggestion
1. Create a reader to read both `MapFile` with a new path and the `configuration` of the filesystem.
```

File: site2/docs/tiered-storage-filesystem.md [hunk 8]
```diff
@@ -520,4 +520,66 @@ Execute the following commands in the repository where you download Pulsar tarba
...
+  ```shell
+     MapFile.Reader reader = new MapFile.Reader(new Path(dataFilePath),  configuration); 
+  ```
+* Read data as `LedgerEntry` from FileSystem.
```
Comment (momo-jun) [inline site2/docs/tiered-storage-filesystem.md:547]:
```suggestion
2. Read the data as `LedgerEntry` from the filesystem.
```

File: site2/docs/tiered-storage-filesystem.md [hunk 9]
```diff
@@ -520,4 +520,66 @@ Execute the following commands in the repository where you download Pulsar tarba
...
+     key.set(nextExpectedId - 1);
+     reader.seek(key);
+     reader.next(key, value);
+     int length = value.getLength();
+     long entryId = key.get();
+     ByteBuf buf = PooledByteBufAllocator.DEFAULT.buffer(length, length);
+     buf.writeBytes(value.copyBytes());
+     LedgerEntryImpl ledgerEntry = LedgerEntryImpl.create(ledgerId, entryId, length, buf);
+  ```
+* Deserialize the `ledgerEntry` to `Message`.
```
Comment (momo-jun) [inline site2/docs/tiered-storage-filesystem.md:560]:
```suggestion
3. Deserialize the `ledgerEntry` to `Message`.
```

Which one is correct? `LedgerEntry` or `ledgerEntry`? The difference between step2 and step3 may make users confused.

File: site2/docs/tiered-storage-filesystem.md [hunk 10]
```diff
@@ -520,4 +520,66 @@ Execute the following commands in the repository where you download Pulsar tarba
...
+
+## Read offloaded data from fileSystem
+
+This section provides detailed instructions about how to read data out as Ledger Entry in the file system.
+
+* The data was offloaded as `MapFile` to the following path:
+  ```properties
+    path = storageBasePath + "/" + managedLedgerName + "/" + ledgerId + "-" + uuid.toString();
+  ```
+    1. storageBasePath is the value of `hadoop.tmp.dir`, configured in `broker.conf` or `filesystem_offload_core_site.xml`
```
Comment (momo-jun) [inline site2/docs/tiered-storage-filesystem.md:533]:
```suggestion
* `storageBasePath` is the value of `hadoop.tmp.dir`, which is configured in `broker.conf` or `filesystem_offload_core_site.xml`.
```

File: site2/docs/tiered-storage-filesystem.md [hunk 11]
```diff
@@ -520,4 +520,66 @@ Execute the following commands in the repository where you download Pulsar tarba
...
+    2. managedLedgerName is the name of the persistentTopic manager Ledger
```
Comment (momo-jun) [inline site2/docs/tiered-storage-filesystem.md:534]: * `managedLedgerName` is the ledger name of the persistentTopic manager.

File: site2/docs/tiered-storage-filesystem.md [hunk 12]
```diff
@@ -520,4 +520,66 @@ Execute the following commands in the repository where you download Pulsar tarba
...
+  ```shell
+     managedLedgerName of persistent://public/default/topics-name is public/default/persistent/topics-name.
+  ```
+  Can use the following method to get the managedLedgerName:
```
Comment (momo-jun) [inline site2/docs/tiered-storage-filesystem.md:538]:
```suggestion
  You can use the following method to get `managedLedgerName`:
```

File: site2/docs/tiered-storage-filesystem.md [hunk 13]
```diff
@@ -520,4 +520,66 @@ Execute the following commands in the repository where you download Pulsar tarba
 
     And the **Capacity Used** is changed from 4 KB to 116.46 KB.
 
-    ![](assets/FileSystem-8.png)
\ No newline at end of file
+    ![](assets/FileSystem-8.png)
+
+## Read offloaded data from fileSystem
```
Comment (momo-jun) [inline site2/docs/tiered-storage-filesystem.md:525]:
```suggestion
## Read offloaded data from filesystem
```

File: site2/docs/tiered-storage-filesystem.md [hunk 14]
```diff
@@ -520,4 +520,66 @@ Execute the following commands in the repository where you download Pulsar tarba
...
+
+This section provides detailed instructions about how to read data out as Ledger Entry in the file system.
```
Comment (momo-jun) [inline site2/docs/tiered-storage-filesystem.md:527]: This sentence can be moved.

**Comment by momo-jun** [comment]
@liangyepianzhou I proposed some suggestions to make the instructions easier to read and understand by changing the ordered/unordered list plus other minor fixes. To provide a comprehensive understanding, you can preview my suggestions as follows. 

The offloaded data is stored as `MapFile` in the following new path of the filesystem:
…
  * `storageBasePath` is the value of `hadoop.tmp.dir`, which is configured in `broker.conf` or `filesystem_offload_core_site.xml`.
  * `managedLedgerName` is the ledger name of the persistentTopic manager. 
      ...
     You can use the following method to get `managedLedgerName`.
      …

To read data out as ledger entries from the filesystem, complete the following steps.
1. Create a reader to read both `MapFile`  with a new path and the `configuration` of the filesystem.
    ...
2. Read the data as `LedgerEntry` from the filesystem.
    ...
3. Deserialize the `ledgerEntry` to `Message`.
    ...

Note that the capitalization of `LedgerEntry` in step2 and `ledgerEntry` in step3 is different, which may confuse users.

**Review by momo-jun — APPROVED** [review]
LGTM.
//cc @Anonymitaet

**Pull request merged**